specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1098 @@
+
+- __Invalid state transition__: Any malicious majority of a subnet can also push invalid state transitions in a valid block. Malicious validators can include messages in a valid block that trigger unallowed state changes. Invalid state transitions are a general misbehavior that may impact HC-specific processes like the commitment of checkpoints or the result of atomic executions.
+In order to report an invalid state transition in a subnet, users need to submit to the parent SCA the chain of blocks from the closest valid committed checkpoint to the block where the invalid state transition happens. The parent SCA replays every state change in the reported chain and verifies that, indeed, an invalid state transition happened in the reported block. If the misbehavior is reported successfully, the collateral for the subnet is correspondingly slashed.
+Thus, if an invalid state transition happens at block `n`, and the latest committed checkpoint was at block `n-x`, a user looking to report the misbheavior needes to submit to the SCA of the parent the chain of `x` blocks from `n-x` to `n` so the state can be replayed and the invalid state transition checked.
+>Note: Along with the checkpoint at `n-x` a snapshot of the state at `n-x` may also need to be provided to enable the whole state of the subnet (for every contract) to be replayed.
+
+## Cross-net routing gas price
+> WIP. This is just a placeholder for this section. The content included is not final. Follow all design updates [here](https://hackmd.io/HpxNIaacTn2_t6jaDs9RcQ)
+
+Cross-net messages from a source subnet, `sn_src`, to some other subnet in the hierarchy,`sn_dest`, need to be provided with enough gas for the message to be executed and routed successfully by every subnet in the path `Path(sn_src, sn_dest)`. Cross-net messages within a subnet are treated like any other message, and their execution triggers some state change that costs a certain amount of gas in the subnet, `gas_cost_sn`. The main difference for cross-net messages is that this execution and state change may translate into the propagation of the message to the next subnet in the path. However, for a user to be able to provide the message with enough gas, it needs to have a sense of the cost of execution in each subnet in the path.
+
+To improve the UX and make cross-net gas costs more predictable, each subnet publishes basic parameters of their gas model at spawning time: 
+- `gas_model_type` (if based on miner tipping, EIP1559-like, etc.)
+- `curr_base_fee` (if any)
+- etc.
+
+With this information, cross-net message will include an array of `gas_limit` amounts that the message is willing to allocate to each subnet in the hop. Subnets in the path won't be able to charge over the specified `gas_limit` for the subnet. When a cross-net reaches a subnet for which `gas_limit` is insufficient, it fails and an error message is propagated back to the source, along with the unspent gas.
```

Comment:
> provision in the message value + sum(gas_limit_subnet)

I'm not familiar with the Filecoin gas model, but in Ethereum you have a gas_limit and also a gas_price (in gwei), with the fees being the product of the two. 

That suggests you need not just an _array of `gas_limit`_ but potentially an array of `gas_price` as well, and you would provision `value + sum(gas_limit * gas_price for (gas_limit, gas_price) in subnet_fee_limits)` in the source.

Reply:
It would be worth mentioning what happens to over-provisioned gas in the happy path, not just when a message fails. Does the sender get it back in the final destination, or do they trickle back to the source? If the latter, at what cost?